### PR TITLE
build and upload docs from circleci

### PR DESCRIPTION
**Pull request by mattip**
Adapted from pytorch/audio#1033 and subsequent fixes. The idea is to build docs on every PR. On nightly and on tag runs, the built docs will be pushed to the gh-pages branch.

### CODE REVIEW

File: .circleci/build_docs/commit_docs.sh [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+    rm -rf ./_static/*
+    cp -r "${src}/docs/build/html/_static/"* ./_static
+    git add --all ./_static || true
+fi
+git add --all ./"${target}" || true
+git config user.email "soumith+bot@pytorch.org"
+git config user.name "pytorchbot"
+# If there aren't changes, don't make a commit; push is no-op
+git commit -m "auto-generating sphinx docs" || true
+git remote add https https://github.com/pytorch/audio.git
```
Comment (fmassa) [inline .circleci/build_docs/commit_docs.sh:34]: I believe we need to change this to point to `vision`

**Review by fmassa — COMMENTED** [review]
Thanks for the PR!

Looks like CI failure is related, see https://app.circleci.com/pipelines/github/pytorch/vision/5820/workflows/8009e653-7c94-42fd-ac93-49c0247becc2/jobs/373582 for example

```
.circleci/unittest/linux/scripts/install.sh
.circleci/unittest/linux/scripts/install.sh: line 10: ./conda/bin/conda: No such file or directory
.circleci/unittest/linux/scripts/install.sh: line 11: conda: command not found

Exited with code exit status 127
CircleCI received exit code 127
```

Could you have a look at fixing this?

File: .circleci/build_docs/commit_docs.sh (hunk 1 above)
Comment (mattip) [reply .circleci/build_docs/commit_docs.sh]: fixed

**Comment by mattip** [comment]
Ready for another review. This now uploads the artifacts, you can see them [here](https://378007-73328905-gh.circle-artifacts.com/0/docs/index.html). 

Numpy uses [this github action](https://github.com/numpy/numpy/blob/master/.github/workflows/circleci.yml) to get an extra `ci/circleci: build artifact` CI run, if you click on the details link it takes you right to the index.html page. For instance see the [CI run in this PR](https://github.com/numpy/numpy/pull/18196), scroll to the bottom and there is the run with the `ci/circleci: build artifact` label

**Comment by mattip** [comment]
... but looking at the 64 builds here, it might be harder to find the link on the CI run box than to click through to the circle-ci interface, click on the build_docs, click on the artifacts tab, and choose one.

**Review by fmassa — APPROVED** [review]
Looks great, thanks!

I think it's fine for now for us to have to click on the Artifacts tab in the CircleCI job to get to the rendered page.

**Comment by mattip** [comment]
Thanks. There may be latent problems with the upload script, but it only triggers on "nightly" builds. I will try to remember to look over the next few days.

**Comment by datumbox** [comment]
@mattip just a heads up, @NicolasHug recently sent us a PR #3268 that fixes some of the formatting issues on the docs. Unfortunately we don't have a lint in place to check for important issues like incorrectly names variables names, missing vars etc. Speaking with @fmassa about this, we wonder if that's something that can be added on the nightly doc_build jobs in sphinx?

**Comment by mattip** [comment]
Sphinx can check some of these things. Note how on the [build docs step of the CI run](https://app.circleci.com/pipelines/github/pytorch/vision/5892/workflows/3dd4b080-ddcc-483b-beca-9084adfeed65/jobs/379422) at the end it prints "build succeeded, 37 warnings.". We can turn these into errors, making the build fail. After the basic warnings are cleared up, sphinx has a [nitpicky mode](https://www.sphinx-doc.org/en/master/usage/configuration.html#confval-nitpicky) for things like improper formatting that can also be turned into errors. I would advise using it in the build run and not the nightly run, so PRs that cause new warnings fail before merging.  
Of course that is only practical once the current warnings are fixed.

See pytorch/pytorch#50356, pytorch/pytorch#41335 for similar work in the pytorch repo.

**Comment by datumbox** [comment]
@mattip Thanks for the references. 

> Note how on the [build docs step of the CI run](https://app.circleci.com/pipelines/github/pytorch/vision/5892/workflows/3dd4b080-ddcc-483b-beca-9084adfeed65/jobs/379422) at the end it prints "build succeeded, 37 warnings.".

Looking at the warnings, I noticed that the linter reports many less critical indentation issues but misses some important mistakes on the documentation. Ideally we would like to capture formatting issues like those listed at #3003 and misspelled/missing/incorrect method parameters like the `bboxes` on the example below:

https://github.com/pytorch/vision/blob/767b23ea361c944870c80057f274c15a8475e204/torchvision/utils.py#L141-L157

 Is the specific lint configurable to flag the above? 

> I would advise using it in the build run and not the nightly run, so PRs that cause new warnings fail before merging. Of course that is only practical once the current warnings are fixed. See pytorch/pytorch#50356, pytorch/pytorch#41335 for similar work in the pytorch repo

Good to see that PyTorch has already adopted a solution. It's definitely worth exploring using a similar approach. We can bring this for discussion with the team.

**Pull request merged**